src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -747,9 +766,33 @@
   should be copied to."
   ([cacrl :- schema/Str
     localcacrl :- schema/Str]
-    (when (fs/exists? cacrl)
-      (ks/mkdirs! (fs/parent localcacrl))
-      (fs/copy cacrl localcacrl))))
+   (when (not= cacrl localcacrl)
+     (let [max-attempts 5]
+       (when (and (fs/exists? cacrl))
+         (ks/mkdirs! (fs/parent localcacrl))
+         ;; Make up to 'max-attempts' tries to copy the cacrl file to the
+         ;; localcacrl file.  The content of the cacrl file is read into memory
+         ;; and parsed for valid CRL pem content during each attempt.  The content
+         ;; in memory is written to the localcacrl file only if it is valid.  This
+         ;; validation is done to protect against a partially written (which could
+         ;; happen if an asynchronous revocation is in process) or corrupt
+         ;; cacrl file being copied.
+         (loop [attempts-left max-attempts
+                ca-crl-last-modified (fs/mod-time cacrl)]
+           (let [cacrl-as-string (slurp cacrl)]
+             (when-let [write-exception (write-local-cacrl!
+                                         localcacrl
+                                         cacrl-as-string)]
+               (if (zero? attempts-left)
+                 (log/error (i18n/trs "Unable to synchronize crl file {0} to {1}: {2}"
+                                      cacrl localcacrl (.getMessage write-exception)))
+                 (do
+                   (Thread/sleep 1000)
+                   (let [new-ca-crl-last-modified (fs/mod-time cacrl)]
+                     (recur (if (= ca-crl-last-modified new-ca-crl-last-modified)
```

Comment:
It's probably fine to just assume that a later modification event would be triggered in the event that the first set of retries all fail.  I don't know enough about when an OS may decide to trigger the notification so I can't say that would be a bad idea - and it definitely seems simpler.  Seems fine to just go with that for now and make it more sophisticated later only if needed.  Assuming the last modified logic were removed here, seems like it would probably be good to remove it from the [associated tk-webserver-jetty9 changes](https://github.com/puppetlabs/trapperkeeper-webserver-jetty9/blob/c409d7916df418f546e5b352c278236733073b40/java/com/puppetlabs/trapperkeeper/services/webserver/jetty9/utils/InternalSslContextFactory.java#L126-L145) as well, since it basically does the same sort of thing.